File: src/main/java/uk/laxd/dndSimulator/feature/rogue/SneakAttack.java
package uk.laxd.dndSimulator.feature.rogue;

import uk.laxd.dndSimulator.action.MeleeAttackAction;
import uk.laxd.dndSimulator.character.Character;
import uk.laxd.dndSimulator.character.CharacterClass;
import uk.laxd.dndSimulator.dice.Die;
import uk.laxd.dndSimulator.dice.Roll;
import uk.laxd.dndSimulator.equipment.WeaponProperty;
import uk.laxd.dndSimulator.feature.StaticFeature;

public class SneakAttack extends StaticFeature {
    private boolean hasSneakAttacked = false;
    private int diceCount = 0;

    public SneakAttack() {
        super("Sneak Attack");
    }

    @Override
    public void onCreate(Character character) {
        diceCount = new Double(Math.ceil(character.getLevel()/2)).intValue();
    }

    @Override
    public void onDamageRoll(MeleeAttackAction action) {
        if(!(action.getWeapon().hasProperty(WeaponProperty.FINESSE) || action.getWeapon().hasProperty(WeaponProperty.RANGE))) {
            // Sneak attack only applies to finesse/ranged attacks
            return;
        }

        if(this.hasSneakAttacked) {
            // 1 sneak attack per turn
            return;
        }

        if(!action.isWithAdvantage()) {
            // Sneak attack only applies if you have adv
            // TODO: Or an ally within 5 ft of target
            return;
        }

        Roll sneakAttackRoll = new Roll();

        for(int i=0; i<diceCount; i++) {
            sneakAttackRoll.addDie(Die.D6);
        }

        action.addDamageRolls(sneakAttackRoll);

        hasSneakAttacked = true;
    }

    @Override
    public void onTurnEnd(Character character) {
        this.hasSneakAttacked = false;
    }

    @Override
    public CharacterClass getCharacterClassRequired() {
        return CharacterClass.ROGUE;
    }
}

File: app/src/main/kotlin/uk/laxd/dndSimulator/feature/rogue/SneakAttack.kt
package uk.laxd.dndSimulator.feature.rogue

import uk.laxd.dndSimulator.action.AttackAction
import uk.laxd.dndSimulator.feature.StaticFeature
import uk.laxd.dndSimulator.action.MeleeAttackAction
import uk.laxd.dndSimulator.character.Character
import uk.laxd.dndSimulator.equipment.WeaponProperty
import uk.laxd.dndSimulator.dice.Roll
import uk.laxd.dndSimulator.dice.Die
import uk.laxd.dndSimulator.character.CharacterClass
import uk.laxd.dndSimulator.dice.RollType
import kotlin.math.ceil

class SneakAttack : StaticFeature("Sneak Attack") {
    private var hasSneakAttacked = false
    private var diceCount = 0

    override fun onCreate(character: Character) {
        diceCount = ceil((character.level / 2).toDouble()).toInt()
    }

    override fun onDamageRoll(action: AttackAction) {
        if(action !is MeleeAttackAction) {
            return
        }

        if (!(action.weapon.hasProperty(WeaponProperty.FINESSE) || action.weapon.hasProperty(WeaponProperty.RANGE))) {
            // Sneak attack only applies to finesse/ranged attacks
            return
        }
        if (hasSneakAttacked) {
            // 1 sneak attack per turn
            return
        }
        if (action.rollType != RollType.ADVANTAGE) {
            // Sneak attack only applies if you have adv
            // TODO: Or an ally within 5 ft of target
            return
        }

        for (i in 0 until diceCount) {
            action.damageRoll?.addDie(Die.D6)
        }

        hasSneakAttacked = true
    }

    override fun onTurnEnd(character: Character) {
        hasSneakAttacked = false
    }

    override val characterClassRequired: CharacterClass
        get() = CharacterClass.ROGUE
}